// @inrupt/solid-client-authn-core/dist/authenticatedFetch/dpopUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateDpopKeyPair = exports.createDpopHeader = void 0;
const jose_1 = require("jose");
const uuid_1 = require("uuid");
const constant_1 = require("../constant");
function normalizeHTU(audience) {
    const audienceUrl = new URL(audience);
    return new URL(audienceUrl.pathname, audienceUrl.origin).toString();
}
async function createDpopHeader(audience, method, dpopKey) {
    return new jose_1.SignJWT({
        htu: normalizeHTU(audience),
        htm: method.toUpperCase(),
        jti: (0, uuid_1.v4)(),
    })
        .setProtectedHeader({
        alg: constant_1.PREFERRED_SIGNING_ALG[0],
        jwk: dpopKey.publicKey,
        typ: "dpop+jwt",
    })
        .setIssuedAt()
        .sign(dpopKey.privateKey, {});
}
exports.createDpopHeader = createDpopHeader;
async function generateDpopKeyPair() {
    const { privateKey, publicKey } = await (0, jose_1.generateKeyPair)(constant_1.PREFERRED_SIGNING_ALG[0]);
    const dpopKeyPair = {
        privateKey,
        publicKey: await (0, jose_1.exportJWK)(publicKey),
    };
    [dpopKeyPair.publicKey.alg] = constant_1.PREFERRED_SIGNING_ALG;
    return dpopKeyPair;
}
exports.generateDpopKeyPair = generateDpopKeyPair;
